testsuite/tests/typing-address/address.ml

```diff
@@ -0,0 +1,117 @@
+module type S = sig type t [@@repr(address)] end;;
+module F (M : S) : S = M;;
+[%%expect{|
+module type S = sig type t [@@repr(address)] end
+module F : functor (M : S) -> S
+|}];;
+
+(* VALID DECLARATIONS *)
+
+module A = struct
+  (* Abstract types can be addresses *)
+  type t [@@repr(address)]
+
+  (* [@@repr(address)] tag here is unnecessary but valid since t has it *)
+  type s = t [@@repr(address)]
+
+  (* Again, valid alias even without tag *)
+  type r = s
+
+  (** Variant types *)
+  type w = Foo | Bar of int [@@repr(address)]
+
+  (* Mutually recursive declarations work as well *)
+  type p = q [@@repr(address)]
+  and q = T of int
+
+  (* Check subtyping: integers are also addresses *)
+  type u = int [@@repr(address)]
+end;;
+[%%expect{|
+module A :
+  sig
+    type t [@@repr(address)]
+    type s = t [@@repr(address)]
+    type r = s
+    type w = Foo | Bar of int [@@repr(address)]
+    type p = q [@@repr(address)]
+    and q = T of int
+    type u = int [@@repr(address)]
+  end
+|}];;
+
+(* Valid using an explicit signature *)
+module M_valid : S = struct type t = T of int end;;
```

Comment:
You should add an %%expect here, otherwise the two phrases are evaluated together and you get a single error.